internal/dynamic6provider/dynamic_functions_test.go

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package dynamic6provider_test
+
+import (
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-go/tfprotov6"
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+	r "github.com/hashicorp/terraform-plugin-testing/helper/resource"
+	"github.com/hashicorp/terraform-plugin-testing/knownvalue"
+	"github.com/hashicorp/terraform-plugin-testing/statecheck"
+	"github.com/hashicorp/terraform-provider-corner/internal/testing/testprovider"
+	"github.com/hashicorp/terraform-provider-corner/internal/testing/testsdk/providerserver"
+	"github.com/hashicorp/terraform-provider-corner/internal/testing/testsdk/resource"
+)
+
+func Test_Dynamic_UsageInFunctions(t *testing.T) {
+	r.UnitTest(t, r.TestCase{
+		// This test verifies that attributes marked as DynamicPseudoType can still be utilized with built-in Terraform functions if the eventual types match up
```

Comment:
> _Was there particular provider functionality being tested here or was there another reason for this testing?_

This test can probably get removed. I initially wrote it after reading this `cty` documentation for `AllowDynamicType`: https://github.com/zclconf/go-cty/blob/main/docs/functions.md#function-parameters, which we could also test [with Provider Defined Functions](https://github.com/hashicorp/terraform/blob/12aa74b5ac1b0d3db2badf12f149bb0ab596c714/internal/providers/functions.go#L146-L149). I wanted to ensure that there were not any additional considerations for provider developers in regards to practitioners using a provider's dynamic type. 

> _`AllowDynamicType` can be set to true to allow not-yet-typed values to be passed. If not set, calls with a dynamic argument will immediately return cty.DynamicVal, and neither the type-checking function nor the implementation function will be called._

I think I might have misunderstood what `AllowDynamicType` actually does, because all of these functions work fine, despite some by default leaving `AllowDynamicType: false`, like [base64decode](https://github.com/hashicorp/terraform/blob/12aa74b5ac1b0d3db2badf12f149bb0ab596c714/internal/lang/funcs/encoding.go#L21).

Reply:
That's great context, thank you. It's certainly confusing, but I _believe_ cty's AllowDynamicType is what enables the creation of type-checking functions where the resultant type is unknown, rather than most Terraform usage where configuration is converted into static types (even for dynamic parameters/return types, since they ultimately are represented with statically typed values).

Reply:
Ah, okay then that makes sense, thanks for the explanation. I removed this test against the built-ins in 5c36673

I made a note to add tests for provider defined functions with dynamic input/output 👍🏻